www/source/docs/scaffolding.md

```diff
@@ -0,0 +1,70 @@
+# Scaffolding
+Scaffolding is default implementations of the build phases and runtime hooks for your application. They also provide some default health check hooks where appropriate to ensure your application is functioning reliably. Customized Scaffolding can be created to facilitate re-usability of common patterns in your organization for developing, building, and running your applications.
+
+## Getting Started
+### New Projects
+Simply run `hab plan init` to leverage the scaffolding. Habitat will look in your application code for known project types and add the appropriate scaffolding.
+
+If you would like to specify the desired scaffolding to use, you can run `hab plan init --scaffolding core/java8-maven-tomcat`
+
+The resulting plan will contain all the available callbacks and variables relevant to the scaffolding type.
+
+### Existing Plans
+
+To begin using Scaffolding, you will need to add the appropriate `pkg_scaffolding`.
+
+```bash
+pkg_name="my_java_app"
+....
+scaffolding="core/scaffolding-java8-maven-tomcat"
+```
+
+### Available Scaffolding
+
+* `core/scaffolding-go17`
+* `core/scaffolding-go18`
+* `core/scaffolding-ruby-mri2.2.6-rails-unicorn`
+* `core/scaffolding-ruby-mri2.3.4-rails-unicorn`
+* `core/scaffolding-ruby-mri2.4.0-rails-unicorn`
+* `core/scaffolding-java8-maven-tomcat`
+* `core/scaffolding-java8-maven-jboss`
+* `core/scaffolding-java8-gradle-tomcat`
+* `core/scaffolding-java8-gradle-jboss`
+* `core/scaffolding-java7-ant`
+* `core/scaffolding-python-unicorn`
+
+## Callbacks
+Each scaffolding defines a set of callbacks which are unique to the scaffolding type. This allows you to complex projects which may need to leverage different scaffoldings as needed within your project. Each of these callbacks define overrides of base level callbacks.
+
+### scaffolding_{flavor}_begin
+Each scaffolding contains a begin callback which is executed immediately while the `scaffolding.sh` file is loaded.
```

Comment:
Yeah, I'd agree that a function would lead to easier testing. Something we've tried to do with core plans is to not have any adverse side effects when a `plan.sh` is sourced in other than setting the package/build-related variables (we use this when [computing](https://github.com/habitat-sh/core-plans/blob/b10d18da313a44679ea142389b1bbb13c6881807/bin/build-dependent-order.rb#L69-L81) dependent package rebuilds). I'd rather have an explicit loading function so we know what "bucket" any code execution should belong in--if it turns out there is no appropriate bucket then maybe we've missed a build phase or an internal setup event.